### DISCUSSION TITLE
Dependecies overrides are not taken into account into the apenapi document

### DISCUSSION THREAD

**Discussion by lmignon** [fastapi/fastapi#8392] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Depends, FastAPI, Header
from fastapi.security import OAuth2PasswordBearer
from fastapi.testclient import TestClient

app = FastAPI()


def get_user_id() -> int:
    pass


def get_user(user_id=Depends(get_user_id)) -> str:
    return f"{user_id}"


oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")


def get_user_id_from_auth_override(token: str = Depends(oauth2_scheme)) -> int:
    return 1


def get_user_id_from_header_override(user_id: int = Header()) -> int:
    return user_id


@app.get("/user")
def read_user(
    user: str = Depends(get_user),
):
    return user


client = TestClient(app)


override_with_security_schema = {
    "components": {
        "securitySchemes": {
            "OAuth2PasswordBearer": {
                "flows": {"password": {"scopes": {}, "tokenUrl": "token"}},
                "type": "oauth2",
            }
        }
    },
    "info": {"title": "FastAPI", "version": "0.1.0"},
    "openapi": "3.0.2",
    "paths": {
        "/user": {
            "get": {
                "operationId": "read_user_user_get",
                "responses": {
                    "200": {
                        "content": {"application/json": {"schema": {}}},
                        "description": "Successful " "Response",
                    }
                },
                "security": [{"OAuth2PasswordBearer": []}],
                "summary": "Read User",
            }
        }
    },
}

app.dependency_overrides[get_user_id] = get_user_id_from_auth_override
app.openapi_schema = None
response = client.get("/openapi.json")
assert response.status_code == 200, response.text
assert response.json() == override_with_security_schema
```

### Description

DI is a great feature. We use it to develop reusable APIs routers packaged in different python modules. Depending of the customer requirements, the routers are assembled into a dedicated App .Each service requiring to get the user information declares a `Depends` on a generic method `get_user`. The `dependencies_override` functionality is then used to define the concrete implementation of this method when the application is assembled according to the customer needs. If the override work as expected in such a case, unfortunately the overrides are not taken into account into the openapi documentation.

### Operating System

Linux

### FastAPI Version

0.85.0

### Python Version

3.10

**Comment by lmignon** [comment]
Any news about this proposal? It's very useful and allows to build reusable routers into different app....

**Comment by YuriiMotov** [comment, marked as answer]
I think it works as it should work.
You shouldn't use `dependencies_override` to change openapi schema. It's only for mocking some dependencies in tests

**Reply by lmignon** [reply]
@YuriiMotov 
We don't use-it to only to change the schema but to be able to create reusable components. With this approach we design for example a library providing a set of routes which depends on the authenticated user and when these routes are mounted into the app, we provide the real implementation as an override of the method specified into the `depends` declaration. It's a kind of dependency injection that allows you to reuse the same library into different contexts. It's very very useful and works in production for years now. The only issue is that the overrides are not taken into account when the openapi schema is generated from the final app. Can you reconsider your response or do I need to find my proper way to do it? I think that this use case is an advanced and powerful feature provided by fastapi.

**Reply by lmignon** [reply]
The related PR is https://github.com/fastapi/fastapi/pull/5452

**Reply by YuriiMotov** [reply]
My opinion is that this is a misuse of `dependencies_override` mechanism.

> This is for testing, to replace expensive dependencies with testing versions.

Source: https://fastapi.tiangolo.com/reference/fastapi/#fastapi.FastAPI.dependency_overrides

It's intended to be used for overriding dependencies in tests. I don't see a point in testing openapi schema with overridden dependencies.

If think you'd better describe your use case with examples and open new discussion with feature request.
If it has wide community support, FastAPI team will take a look and probably accept this feature request

**Reply by YuriiMotov** [reply]
Good explanation of the idea in this comment: https://github.com/fastapi/fastapi/pull/5452#issuecomment-2967610089 (just added the link for people who will read this)
